### ISSUE TITLE
Style/StringLiteralsInInterpolation probably shouldn’t trigger inside a heredoc

### ISSUE THREAD

**Issue by brandonweiss** [rubocop-hq/rubocop#4482]
The point of `Style/StringLiteralsInInterpolation` is presumably to prevent weirdness like this:

```ruby
"Tests #{success ? "PASS" : "FAIL"}"
```

So it enforces something like this:

```ruby
"Tests success ? 'PASS' : 'FAIL'"
```

But right now it's triggering for something like this:

```ruby
<<~EOS
  Tests #{success ? "PASS" : "FAIL"}
EOS
```

Inside of a heredoc the quotes inside an interpolation shouldn't matter and it should force whatever quote style `Style/StringLiterals` is set to.

**Comment by bbatsov** [comment]
> The point of Style/StringLiteralsInInterpolation is presumably to prevent weirdness like this:

It's actually to prevent redundant interpolation of literals, as you can just write those as is. The ternary op shouldn't trigger this cop.

**Comment by brandonweiss** [comment]
Ohhh I didn't realize. That also makes sense to me, though! It might have to handle a few cases, though. I've also seen things like this:

```ruby
<<~EOS
  <div class="#{'active' if active?}">
  </div>
EOS
```

I imagine it shouldn't trigger for that either.

**Comment by bbatsov** [comment]
Yeah, it shouldn't.

**Comment by smakagon** [comment]
Description for this cop says:
> Checks if uses of quotes match the configured preference.

So if I prefer single quotes but I have interpolation with double quoted strings inside:
```
x = 10
foo = "hello: #{x < 10 ? "foo" : "bar"}"
```
It will return:
> Prefer single-quoted strings inside interpolations.

To fix this issue, should we ignore it just for heredoc?
It seems like a good idea to have this check for ternary operators.

**Comment by bbatsov** [comment]
Hmm, I'll defer this to @jonas054. As the cop doesn't have much documentation it's a bit unclear to me what was it supposed to do exactly. But upon further inspection it seems that @brandonweiss was right - I simply misinterpreted what the cop was doing based on its name.

**Comment by smakagon** [comment]
Yes, documentation should be extended to at least show some examples. 
I had to go through the source code to figure out how it works. 
But code is pretty good and shows the intent:
```
        def offense?(node)
          # If it's not a string within an interpolation, then it's not an
          # offense for this cop.
          return false unless inside_interpolation?(node)

          wrong_quotes?(node)
        end
```

All it does, it checks that it's interpolation and code inside interpolation follows the same conventions regarding quotes (double or single).

**Comment by smakagon** [comment]
@bbatsov @brandonweiss created PR https://github.com/bbatsov/rubocop/pull/4527 which extends documentation for this cop. At least that should help to figure out what this cop does.

**Comment by jonas054** [comment]
As @brandonweiss remarked, `Style/StringLiteralsInInterpolation` does two things.
1. If `EnforcedStyle` is `single_quotes`, prevents unreadable code like
```ruby
"#{success ? "pass#{score > 50 ? " with flying colors" : ""}" : "fail"}"
```
Yes. That's legal. :smile: 

2. Enforces a preferred quoting style for interpolated strings, whether the code has readability problems (like 1.) or not (like heredocs).

My suggestion is to add a `Lint` cop, e.g. `Lint/NestedDoubleQuotes` that deals only with case 1, and let `Style/StringLiteralsInInterpolation` deal only with case 2.

**Comment by bbatsov** [comment]
> My suggestion is to add a Lint cop, e.g. Lint/NestedDoubleQuotes that deals only with case 1, and let Style/StringLiteralsInInterpolation deal only with case 2.

Sounds good to me!

**Comment by brandonweiss** [comment]
OK, I had a go at this in https://github.com/bbatsov/rubocop/pull/4613!
